docs/implementation.html

```diff
@@ -78,7 +78,49 @@ <h5><a id="controlbatch" href="#controlbatch">5.3.1.1 Control Batches</a></h5>
     </pre></p>
     <p>The schema for the value of a control record is dependent on the type. The value is opaque to clients.</p>
 
-	<h4><a id="record" href="#record">5.3.2 Record</a></h4>
+    <h5><a id="messageset" href="#messageset">5.3.1.2 Old Message Format</a></h5>
+    <p>
+        Prior to Kafka 0.11 messages were transferred and stored in a format called MessageSets. In MessageSets each message has its own metadata. Note that although MessageSets are formally arrays,
+        they are not preceded by an int32 array size like other array elements in the protocol.
+    </p>
+
+    <b>Message Set:</b><br>
+    <p><pre class="brush: java;">
+    MessageSet (Version: 0) => [offset message_size message]
+        offset => INT64
+        message_size => INT32
+        message => crc magic_byte attributes key value
+            crc => INT32
+            magic_byte => INT8
+            attributes => INT8
+            key => BYTES
+            value => BYTES
+    </pre></p>
+    <p><pre class="brush: java;">
+    MessageSet (Version: 1) => [offset message_size message]
+        offset => INT64
+        message_size => INT32
+        message => crc magic_byte attributes key value
+            crc => INT32
+            magic_byte => INT8
+            attributes => INT8
+            timestamp =>INT64
+            key => BYTES
+            value => BYTES
+    </pre></p>
+    <p>
+        When the producer is sending non compressed messages, it can set the offsets to anything.
+        When the producer is sending compressed messages, to avoid server side re-compression, each compressed message should have offset starting from 0 and increasing by one for each inner message in the compressed message.
+    </p>
+    <p>
+        In versions until Kafka 0.10 the magic byte value was 0. Value 1 i.e. message format version 1 was introduced with timestamp support in version 0.10 .
+        In version 1 the fourth lowest bit of attributes byte represents the timestamp type. 0 stands for CreateTime and 1 stands for LogAppendTime.
+        The producer should always set this bit to 0. Similarly to version 2 above, the lowest bits of attributes mean compression type and the highest bits must be set to 0.
+    </p>
+    <p>In message format versions 0 and 1 Kafka supports recursive messages in which case the value field may itself contain a message set.</p>
```

Comment:
We should probably mention that this is only for compressed message sets. Also, maybe this can go above the comment above about compression which refers to "inner messages."